packages/jupyterlab-manager/src/semverregistry.ts

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import { maxSatisfying } from 'semver';
+
+import { ArrayExt } from '@lumino/algorithm';
+
+import { PromiseDelegate } from '@lumino/coreutils';
+
+import { Throttler } from '@lumino/polling';
+
+const TIMEOUT = 300;
+
+interface IRequest<T> {
+  name: string;
+  semVer: string;
+  result: PromiseDelegate<T | undefined>;
+  timestamp: number | null;
+}
+
+/**
+ * A registry using semver ranges to retrieve values.
+ *
+ * Each object in the registry is associated with a name and a semantic
+ * version. Objects can be retrieved by giving the name and a semantic version
+ * range, and the maximum satisfying version of the object will be returned.
+ *
+ * Object retrieval may be delayed for a short timeout if no matching object
+ * can be found. This is to allow a short window for objects to be registered
+ * before get requests are resolved as unsatisfied. Practically, this means
+ * that initial object registration can happen concurrent with the first
+ * requests.
+ *
+ */
+export class SemVerRegistry<T> {
+  /**
+   * Register a object with a given name and version.
+   */
+  register(options: SemVerRegistry.IRegistryData<T>): void {
+    const { name, version, object } = options;
+    let versions = this._registry.get(name);
+    if (versions === undefined) {
+      versions = new Map<string, T>();
+      this._registry.set(name, versions);
+    }
+
+    if (versions.has(version)) {
+      throw new Error(`Version ${version} of ${name} already registered.`);
+    }
+    versions.set(version, object);
+
+    // Some pending requests may now be satisfied, so schedule a retry.
+    if (this._pendingRequests.length > 0) {
+      this._throttle.invoke();
+    }
+  }
+
+  /**
+   * Get an object for a name and semVer range.
+   *
+   * @returns a promise resolving to the object with the maximum version
+   * satisfying the version range if one is registered, or undefined if one is
+   * not registered.
+   *
+   * #### Notes
+   * Resolution may be delayed for a short timeout to allow for some delay in
+   * registering the object.
+   */
+  async get(name: string, semVer: string): Promise<T | undefined> {
+    const mod = this._getSync(name, semVer);
+    if (mod) {
+      return mod;
+    }
+    const result = new PromiseDelegate<T | undefined>();
+    const timestamp = Date.now();
+    this._pendingRequests.push({ name, semVer, result, timestamp });
```

Comment:
@jasongrout do you think we need to do something here for the case where you don't have e.g. bqplot, and it's never going to load? In that case, we currently might delay the notebook execution by 300ms every time a widget is shown (while IMO we shoudl only delay it the first time bqplot is requested).

Reply:
Well, we only delay responses (which I think you are right, will delay notebook execution, fair enough) when you request a widget that is not there. If a widget is shown that is there, there is no delay (other than an async promise that immediately resolves).

If there is a complicated bqplot object, than we might have 20 bqplot models that will be requested right off. We can't delay just the first and error out the other 19, because if bqplot does load, we want to resolve all 20 correctly. What are you imagining should happen in this case?

Reply:
The issue would be that we also delay notebook execution when interactively executing the notebook. I'll give this a shot + fix it if it's actually an issue.

Reply:
>  I'll give this a shot + fix it if it's actually an issue.

We're getting closer to needing to release 8.0, so either we need to move this PR forward or bump it to a later release.

Reply:
Hi @jasongrout I just checked again with four bqplot widgets in a notebook. It remains "snappy" so I don't think our current logic is an issue. Looks good to merge to me!

Reply:
Ok turns out I was hitting some cache which made it load the old version or something like that.

It doesn't work at all right now, as it's impossible to `invoke()` the throttler when it's in `invoked` state, meaning the throttler only runs exactly _once_ right now, not enough to clear the timeouts.

I'll look into either switching to the poller in Lumino ...